### PR TITLE
[ci skip] Create a new RequestHeader and made it a part of KVRequestHeader

### PR DESCRIPTION
**Pull request by vivekmenezes**
RFC: This code compiles but the tests still need some updates. I'd like to know whether I should take it to the end.

This is to make Request/Response/Call work well with the structured data API allowing us to reuse code to its maximum.

See the changes in proto/api.proto and see its effect on say kv/txn_coord_sender.go

Thanks

### CODE REVIEW

File: proto/api.proto [hunk 1]
```diff
@@ -135,27 +119,53 @@ message ResponseHeader {
...
+
+// KVResponseHeader is supplied with every KV node request.
+message KVResponseHeader {
+  optional ResponseHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+}
+
 // A GetRequest is the argument for the Get() method.
 message GetRequest {
-  optional RequestHeader header = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
+  optional KVRequestHeader kvheader = 1 [(gogoproto.nullable) = false, (gogoproto.embed) = true];
```
Comment (petermattis) [inline proto/api.proto]: Why did you change the name of the field? It should be fine to leave it named `header`.

File: proto/api.go [hunk 2]
```diff
@@ -79,6 +79,7 @@ type Request interface {
 	gogoproto.Message
 	// Header returns the request header.
 	Header() *RequestHeader
+	KVHeader() *KVRequestHeader
```
Comment (petermattis) [inline proto/api.go]: If you're adding `KVHeader()` here, then the interface should be `KVRequest`. I think this should really be split up to be something like:

```
type Request interface {
  gogoproto.Message
  Header() *RequestHeader
  Method() Method
  flags() int
}

type KVRequest interface {
  Request
  KVHeader() *KVRequestHeader
  CreateReply() KVResponse
}
```

File: proto/api.proto (hunk 1 above)
Comment (petermattis) [inline proto/api.proto]: Due to the `gogoproto.embed` option, the field name isn't even used. Definitely leave as `header`.

File: proto/api.go (hunk 2 above)
Comment (vivekmenezes) [inline proto/api.go]: But Call uses Request and Response. And any sender depends on Call.

File: proto/api.proto (hunk 1 above)
Comment (vivekmenezes) [inline proto/api.proto]: sounds good will make that change

File: proto/api.go (hunk 2 above)
Comment (petermattis) [inline proto/api.go]: If `Call` is still going to be used everywhere, the `Request` and `Response` it contains cannot be KV specific or you haven't bought anything with this change.

Comment (vivekmenezes) [inline proto/api.go]: Structured would use Request/Response with KVHeader() returning nil. That was my plan.

Comment (petermattis) [inline proto/api.go]: Hmm. That's kind of ugly. I don't have a better solution to offer, though. You should add a comment explaining this, otherwise it is surprising to the reader that `Request` has a `KVHeader()` method.

Comment (petermattis) [inline proto/api.go]: There are only 27 references to the `Key`, `EndKey`, `Replica` and `RaftID` fields of `RequestHeader`. I think it would be reasonable to fix these uses to operation on a `KVRequest` instead of a `Request`. The majority are in `dist_sender.go` which isn't going to be able to send non-KV requests anyways.

Comment (bdarnell) [inline proto/api.go]: Instead of KVHeader returning nil for structured requests, I'd rather split the interfaces as Pete described and have the senders that need KV fields (just DistSender?) type-assert the Request to a KVRequest. This is consistent with how a lot of the stdlib works (i.e. "this method takes an io.Reader. If it is also an io.ReadCloser it will be closed as well").

File: proto/api.proto (hunk 1 above)
Comment (bdarnell) [inline proto/api.proto]: I dislike relying on the embed option here. In go, we can use request.Header() and pretend the intermediate KVHeader doesn't exist, but in other languages you'll have to go through the intermediate `{KV,Structured}Header`, and that will make it more difficult to deal with all requests in a generic way. I would prefer to have two separate top-level fields for RequestHeader and KVRequestHeader.

File: proto/api.go (hunk 2 above)
Comment (bdarnell) [inline proto/api.go]: Also note that the `Sender` interface used to be named `KVSender`. We could bring that back (along with a `KVCall`) and separate the interfaces implemented by the different senders (requiring KVRequests in DistSender and LocalSender but any Request in HTTPSender).

File: proto/api.proto (hunk 1 above)
Comment (spencerkimball) [inline proto/api.proto]: No other clients or languages should ever have to deal with these headers in just a few weeks.

File: proto/api.go (hunk 2 above)
Comment (vivekmenezes) [inline proto/api.go]: I'll go ahead and split that interface and see where it take me. Thanks for pointing me to the stdlib

File: proto/api.proto (hunk 1 above)
Comment (bdarnell) [inline proto/api.proto]: For KV, yes, but the structured protos will be used by all the clients.

Comment (spencerkimball) [inline proto/api.proto]: Yeah, probably shouldn't use any of the particularly fancy options in what becomes our public .proto file.

Comment (vivekmenezes) [inline proto/api.proto]: Hey after talking with Pete I'm going to table this PR because its not
clear what the new api is going to look like. I'll get back to this if we
really need it. In other words we've agreed that all this stuff is internal
for now.

**Comment by tschottdorf** [comment]
heads up: the `[ci skip]` is better kept at the end of the commit message (as opposed to the title).
